File: qc-application/src/main/java/com/eager/questioncloud/application/business/payment/question/implement/QuestionPaymentCouponReader.java
package com.eager.questioncloud.application.business.payment.question.implement;

import com.eager.questioncloud.core.domain.coupon.infrastructure.repository.CouponRepository;
import com.eager.questioncloud.core.domain.coupon.infrastructure.repository.UserCouponRepository;
import com.eager.questioncloud.core.domain.coupon.model.Coupon;
import com.eager.questioncloud.core.domain.coupon.model.UserCoupon;
import com.eager.questioncloud.core.domain.payment.model.QuestionPaymentCoupon;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class QuestionPaymentCouponReader {
    private final UserCouponRepository userCouponRepository;
    private final CouponRepository couponRepository;

    public QuestionPaymentCoupon getQuestionPaymentCoupon(Long userCouponId, Long userId) {
        if (userCouponId == null) {
            return QuestionPaymentCoupon.noCoupon();
        }

        UserCoupon userCoupon = userCouponRepository.getUserCoupon(userCouponId, userId);
        userCoupon.validate();

        Coupon coupon = couponRepository.findById(userCoupon.getCouponId());

        return QuestionPaymentCoupon.create(userCouponId, coupon);
    }
}

File: qc-application/src/main/java/com/eager/questioncloud/application/business/payment/question/implement/QuestionPaymentCouponReader.kt
package com.eager.questioncloud.application.business.payment.question.implement

import com.eager.questioncloud.core.domain.coupon.infrastructure.repository.CouponRepository
import com.eager.questioncloud.core.domain.coupon.infrastructure.repository.UserCouponRepository
import com.eager.questioncloud.core.domain.payment.model.QuestionPaymentCoupon
import com.eager.questioncloud.core.domain.payment.model.QuestionPaymentCoupon.Companion.create
import org.springframework.stereotype.Component

@Component
class QuestionPaymentCouponReader(
    private val userCouponRepository: UserCouponRepository,
    private val couponRepository: CouponRepository,
) {
    fun getQuestionPaymentCoupon(userCouponId: Long?, userId: Long): QuestionPaymentCoupon? {
        if (userCouponId == null) {
            return null
        }

        val userCoupon = userCouponRepository.getUserCoupon(userCouponId, userId)
        userCoupon.validate()

        val coupon = couponRepository.findById(userCoupon.couponId)

        return create(userCouponId, coupon)
    }
}
